### PR TITLE
VTOrc converts a tablet to DRAINED type if it detects errant GTIDs on it

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

This PR introduces a new VTOrc flag `--change-tablets-with-errant-gtid-to-drained` that allows the users to toggle the ability of VTOrc to change the tablet type of tablets with errant GTIDs to `DRAINED`. By default, the flag is false.

This feature allows the users to configure VTOrc such that any tablet that encounters errant GTIDs is automatically taken out of the serving graph. These tablets can then be inspected for what the errant GTIDs are, and once fixed, they can rejoin the cluster.

When VTOrc changes the tablet to a DRAINED type, it also forgets it from its own datastore, and the user has to change the tablet back to its original type for VTOrc to start monitoring it again.
## Related Issue(s)

- Fixes https://github.com/vitessio/vitess/issues/13872

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation was added or is not required - https://github.com/vitessio/website/pull/1574

### CODE REVIEW

File: go/vt/vtorc/config/config.go [hunk 1]
```diff
@@ -82,6 +83,7 @@ func RegisterFlags(fs *pflag.FlagSet) {
 	fs.DurationVar(&topoInformationRefreshDuration, "topo-information-refresh-duration", topoInformationRefreshDuration, "Timer duration on which VTOrc refreshes the keyspace and vttablet records from the topology server")
 	fs.DurationVar(&recoveryPollDuration, "recovery-poll-duration", recoveryPollDuration, "Timer duration on which VTOrc polls its database to run a recovery")
 	fs.BoolVar(&ersEnabled, "allow-emergency-reparent", ersEnabled, "Whether VTOrc should be allowed to run emergency reparent operation when it detects a dead primary")
+	fs.BoolVar(&convertTabletWithErrantGTIDs, "convert-tablets-with-errant-gtids", convertTabletWithErrantGTIDs, "Whether VTOrc should be changing the type of the tablets with errant GTIDs to DRAINED")
```
Comment (shlomi-noach) [inline go/vt/vtorc/config/config.go:86]: This variable doesn't seem to be actually used? The function `ConvertTabletWithErrantGTIDs()` just returns `ersUsed`.

File: go/vt/vtorc/inst/analysis_dao.go [hunk 2]
```diff
@@ -475,6 +477,9 @@ func GetReplicationAnalysis(keyspace string, shard string, hints *ReplicationAna
 			a.Analysis = PrimarySemiSyncMustNotBeSet
 			a.Description = "Primary semi-sync must not be set"
 			//
+		} else if topo.IsReplicaType(a.TabletType) && a.ErrantGTID != "" && config.ConvertTabletWithErrantGTIDs() {
```
Comment (shlomi-noach) [inline go/vt/vtorc/inst/analysis_dao.go:480]: The detection should not be conditioned by the course of action. Analysis should be made for `ErrantGTIDDetected` irrespective of  `config.ConvertTabletWithErrantGTIDs()`.

File: go/vt/vtorc/logic/topology_recovery.go [hunk 3]
```diff
@@ -882,3 +893,38 @@ func fixReplica(ctx context.Context, analysisEntry *inst.ReplicationAnalysis) (r
 	err = setReplicationSource(ctx, analyzedTablet, primaryTablet, reparentutil.IsReplicaSemiSync(durabilityPolicy, primaryTablet, analyzedTablet))
 	return true, topologyRecovery, err
 }
+
+// recoverErrantGTIDDetected changes the tablet type of a replica tablet that has errant GTIDs.
+func recoverErrantGTIDDetected(ctx context.Context, analysisEntry *inst.ReplicationAnalysis) (recoveryAttempted bool, topologyRecovery *TopologyRecovery, err error) {
```
Comment (shlomi-noach) [inline go/vt/vtorc/logic/topology_recovery.go:898]: Here (in this function) is where we want to check `config.ConvertTabletWithErrantGTIDs()`

**Review by shlomi-noach — CHANGES_REQUESTED** [review]

File: go/vt/vtorc/config/config.go (hunk 1 above)
Comment (GuptaManan100) [reply go/vt/vtorc/config/config.go:86]: Fixed.

File: go/vt/vtorc/inst/analysis_dao.go (hunk 2 above)
Comment (GuptaManan100) [reply go/vt/vtorc/inst/analysis_dao.go]: Fixed.

File: go/vt/vtorc/logic/topology_recovery.go (hunk 3 above)
Comment (GuptaManan100) [reply go/vt/vtorc/logic/topology_recovery.go:902]: Instead done in `getCheckAndRecoverFunctionCode`

**Review by shlomi-noach — APPROVED** [review]

File: changelog/18.0/18.0.0/summary.md [hunk 4]
```diff
@@ -46,6 +47,14 @@ reparent operations. The users that want VTOrc to fix the replication issues, bu
 should start using this flag. By default, VTOrc will be able to run `EmergencyReparentShard`. The users must specify the
 flag to `false` to change the behaviour.
 
+#### <a id="new-flag-errant-gtid-convert"/>VTOrc flag `--convert-tablets-with-errant-gtids`
+
+VTOrc has a new flag `--convert-tablets-with-errant-gtids` that allows the users to toggle the ability of VTOrc to change the
```
Comment (deepthi) [inline changelog/18.0/18.0.0/summary.md:52]:
```suggestion
VTOrc has a new flag `--convert-tablets-with-errant-gtids` that allows users to choose whether VTOrc should change the 
```

File: changelog/18.0/18.0.0/summary.md [hunk 5]
```diff
@@ -46,6 +47,14 @@ reparent operations. The users that want VTOrc to fix the replication issues, bu
 should start using this flag. By default, VTOrc will be able to run `EmergencyReparentShard`. The users must specify the
 flag to `false` to change the behaviour.
 
+#### <a id="new-flag-errant-gtid-convert"/>VTOrc flag `--convert-tablets-with-errant-gtids`
+
+VTOrc has a new flag `--convert-tablets-with-errant-gtids` that allows the users to toggle the ability of VTOrc to change the 
+tablet type of tablets with errant GTIDs to `DRAINED`. By default, the flag is false.
+
+This feature allows the users to configure VTOrc such that any tablet that encounters errant GTIDs is automatically taken out of the
```
Comment (deepthi) [inline changelog/18.0/18.0.0/summary.md:55]:
```suggestion
This feature allows users to configure VTOrc such that any tablet that encounters errant GTIDs is automatically taken out of the
```

File: go/flags/endtoend/vtorc.txt [hunk 6]
```diff
@@ -15,6 +15,7 @@ Usage of vtorc:
       --config-persistence-min-interval duration                    minimum interval between persisting dynamic config changes back to disk (if no change has occurred, nothing is done). (default 1s)
       --config-type string                                          Config file type (omit to infer config type from file extension).
       --consul_auth_static_file string                              JSON File to read the topos/tokens from.
+      --convert-tablets-with-errant-gtids                           Whether VTOrc should be changing the type of the tablets with errant GTIDs to DRAINED
```
Comment (deepthi) [inline go/flags/endtoend/vtorc.txt:18]: I wonder what the best descriptive name for this flag is.
1. disable-tablets-with-errant-gtids
2. drain-tablets-with-errant-gtids - The problem with this is that there is no real "draining" happening here

File: go/vt/vtorc/config/config.go (hunk 1 above)
Comment (deepthi) [inline go/vt/vtorc/config/config.go:86]:
```suggestion
	fs.BoolVar(&convertTabletsWithErrantGTIDs, "convert-tablets-with-errant-gtids", convertTabletWithErrantGTIDs, "Whether VTOrc should be changing the type of tablets with errant GTIDs to DRAINED")
```

File: go/flags/endtoend/vtorc.txt (hunk 6 above)
Comment (shlomi-noach) [reply go/flags/endtoend/vtorc.txt:18]: At the risk of running too long, how about `--change-tablets-with-errant-gtid-to-drained`? It is the most accurate description of the actions taken.

File: changelog/18.0/18.0.0/summary.md (hunk 4 above)
Comment (GuptaManan100) [reply changelog/18.0/18.0.0/summary.md]: Fixed

File: changelog/18.0/18.0.0/summary.md (hunk 5 above)
Comment (GuptaManan100) [reply changelog/18.0/18.0.0/summary.md]: Fixed

File: go/flags/endtoend/vtorc.txt (hunk 6 above)
Comment (GuptaManan100) [reply go/flags/endtoend/vtorc.txt]: I think @shlomi-noach's suggestion is the best we have had so far. I have updated the PR with this name.

File: go/vt/vtorc/config/config.go (hunk 1 above)
Comment (GuptaManan100) [reply go/vt/vtorc/config/config.go]: Fixed

**Review by deepthi — APPROVED** [review]

**Pull request merged**